### PR TITLE
Teams Adapter

### PR DESCRIPTION
**Pull request by bastjan**
## Summary

* Updates the Keycloak Client to handle group hierarchies.
* Reconciles teams by creating, updating and deleting Keycloak groups.
* Periodically imports non-existing groups from Keycloak.

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.

### CODE REVIEW

File: keycloak/client.go [hunk 1]
```diff
@@ -3,16 +3,57 @@ package keycloak
...
+
+// Path returns the path of the group.
+func (g Group) Path() string {
+	if len(g.path) == 0 {
+		return ""
+	}
+	return fmt.Sprintf("/%s", strings.Join(g.path, "/"))
+}
+
+// Path returns the path of the group.
```
Comment (glrf) [inline keycloak/client.go:44]: Wrong comment :spaghetti:

File: controllers/periodic_syncer.go [hunk 2]
```diff
@@ -52,7 +66,42 @@ func (r *OrganizationReconciler) Sync(ctx context.Context) error {
...
+func (r *PeriodicSyncer) syncGroup(ctx context.Context, g keycloak.Group, org *orgv1.Organization) (runtime.Object, error) {
+	const depth = 0
+	switch len(g.PathMembers()) - depth {
+	case 1:
+		return r.syncOrganization(ctx, g, org)
+	case 2:
+		return r.syncTeam(ctx, g)
+	}
+
+	return nil, fmt.Errorf("invalid group hierarchy `%s`", g.Path())
```
Comment (glrf) [inline controllers/periodic_syncer.go:78]: Not sure if we should return an error in that case. I don't think more keycloak sub-groups are invalid, we just don't sync them to the cluster

File: controllers/team_controller.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+// TeamReconciler reconciles a Team object
+type TeamReconciler struct {
+	client.Client
+	Recorder record.EventRecorder
+	Scheme   *runtime.Scheme
+
+	Keycloak KeycloakClient
+
+	// ClusteRoles to give to group members when importing
+	SyncClusterRoles []string
```
Comment (glrf) [inline controllers/team_controller.go:26]: I think that's unused

**Review by glrf — APPROVED** [review]
LGTM :rocket:

File: controllers/periodic_syncer.go (hunk 2 above)
Comment (bastjan) [reply controllers/periodic_syncer.go:78]: I'd rather throw if we don't support something than do nothing...

File: keycloak/client_delete_test.go [hunk 4]
```diff
@@ -27,13 +28,21 @@ func TestDeleteGroup_simple(t *testing.T) {
...
+func newGocloakGroup(id string, path ...string) *gocloak.Group {
+	if len(path) == 0 {
+		panic("group must have at least one element in path")
+	}
+	return &gocloak.Group{
+		ID:   &id,
+		Name: gocloak.StringP(path[len(path)-1]),
+		Path: gocloak.StringP("/" + strings.Join(path, "/")),
+	}
+}
```
[the remark this answers is not in the document]
Comment (bastjan) [reply keycloak/client_delete_test.go:48]: because i used it there first 🙈

File: controllers/periodic_syncer.go (hunk 2 above)
Comment (bastjan) [reply controllers/periodic_syncer.go:78]: But yeah i see your point.

Comment (bastjan) [reply controllers/periodic_syncer.go:78]: I'll log an error but won't abort the sync. I feel like that's more useful here.

**Review by glrf — APPROVED** [review]

**Pull request merged**